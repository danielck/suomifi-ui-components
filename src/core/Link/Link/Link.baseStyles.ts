import { SuomifiTheme } from '../../theme';
import { css } from 'styled-components';
import { baseStyles } from '../BaseLink/BaseLink.baseStyles';
import { MarginProps, getCssMargins } from '../../theme/utils/spacing';

export const LinkStyles = (theme: SuomifiTheme, margins?: MarginProps) => css`
  ${getCssMargins(margins)}
  ${baseStyles(theme)}
`;
