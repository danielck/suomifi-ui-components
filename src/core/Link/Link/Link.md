```js
import { Link, Paragraph } from 'suomifi-ui-components';

<Paragraph>
  <Link
    className="test-classname"
    href="https://www.notvisitedlink.com/"
  >
    Not visited link
  </Link>{' '}
  <Link className="test-classname" href="#">
    Visited link
  </Link>
</Paragraph>;
```

### Skip link

Show link for keyboard navigation.

- Should be the first element that can gain focus on the page with the tab.
- The target container should have `tabIndex={-1}` to handle better the functionality with screenreaders.
- Because of the tabIndex you might want set the style of `outline=none` to the target container.
- To see the component, click below on the dashed area and then press `Tab`.

```js
import { SkipLink } from 'suomifi-ui-components';
<div>
  <div
    style={{
      height: '80px',
      width: '210px',
      borderStyle: 'dashed',
      borderColor: '#000',
      borderWidth: '1px'
    }}
  >
    <SkipLink href="#">Skip to main content</SkipLink>
  </div>
</div>;
```

### External link

```js
import { ExternalLink } from 'suomifi-ui-components';

<>
  <ExternalLink
    href="https://designsystem.suomi.fi/fi/"
    labelNewWindow="Opens to a new window"
  >
    External link opens to new window
  </ExternalLink>
  <ExternalLink
    href="https://designsystem.suomi.fi/fi/"
    toNewWindow={false}
  >
    External link in same window
  </ExternalLink>
</>;
```

### Router link

This component is mainly intended to be used with external libraries/frameworks.

- Renders as the component of your choice, for example <a href="https://reactrouter.com/docs/en/v6/components/link" target="_blank">React Router Link</a>, but with proper Suomi.fi link styles
- Full Typescript support for props based on the component passed to `asComponent`
- Full theme customisation support through `SuomifiThemeProvider`

```js
import { RouterLink } from 'suomifi-ui-components';
import { forwardRef, useRef } from 'react';

const Component = forwardRef((props, ref) => {
  const { children, ...passProps } = props;
  return (
    <a {...passProps} ref={ref}>
      Some component - {children}
    </a>
  );
});

const r = useRef();

<>
  <button onClick={() => console.log(r.current)}>
    Log ref to console
  </button>
  <RouterLink
    href="https://suomi.fi"
    target="_blank"
    asComponent={Component}
    ref={r}
  >
    Testing
  </RouterLink>
</>;
```
