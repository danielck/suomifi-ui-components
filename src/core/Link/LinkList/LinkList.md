A styled list for providing a list of links.

- Use a descriptive heading or a label with the list.
- Link heading or label to the list via the `ariaDescribedBy` prop.
- Use the right heading level semantically and use the `as` prop for changing the styling if needed.
- Wrap each child in `LinkListItem` component for the correct styling.
- `LinkListItem` supports `Link`, `ExternalLink` and `RouterLink` as its children.

```js
import {
  Link,
  ExternalLink,
  LinkListItem,
  RouterLink,
  LinkList,
  Heading
} from 'suomifi-ui-components';

<div>
  <Heading variant="h5" as="h3" id="heading">
    Useful resources
  </Heading>
  <LinkList aria-describedby="heading">
    <LinkListItem>
      <Link href="/">More information on subject X</Link>
    </LinkListItem>
    <LinkListItem>
      <ExternalLink
        href="https://designsystem.suomi.fi/fi/"
        labelNewWindow="Opens to a new window"
      >
        About subject X on a third party site
      </ExternalLink>
    </LinkListItem>
    <LinkListItem>
      <RouterLink
        href="https://suomi.fi"
        target="_blank"
        asComponent="a"
      >
        An additional useful link
      </RouterLink>
    </LinkListItem>
  </LinkList>
</div>;
```
