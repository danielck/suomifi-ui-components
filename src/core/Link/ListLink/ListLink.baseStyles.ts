import { SuomifiTheme } from '../../theme';
import { css } from 'styled-components';

export const ListLinkStyles = (theme: SuomifiTheme) => css`
  &.fi-list-link {
    list-style: none;
    padding: 0;
    & .fi-list-link_icon {
      margin-right: 2px;
      & .fi-icon {
        margin-left: -3px;
        transform: translateY(0.1em);
        & .fi-icon-base-fill {
          fill: ${theme.colors.accentBase};
        }
      }
    }
  }
`;
