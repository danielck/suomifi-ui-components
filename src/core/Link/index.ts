export { Link, LinkProps } from './Link/Link';
export { ExternalLink, ExternalLinkProps } from './ExternalLink/ExternalLink';
export { SkipLink, SkipLinkProps } from './SkipLink/SkipLink';
export { RouterLink, RouterLinkProps } from './RouterLink/RouterLink';
export { LinkListItem, LinkListItemProps } from './LinkListItem/LinkListItem';
export { LinkList, LinkListProps } from './LinkList/LinkList';
