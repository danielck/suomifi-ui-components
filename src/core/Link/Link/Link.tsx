import React, { forwardRef } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { LinkStyles } from '../Link/Link.baseStyles';
import { SuomifiThemeProp, SuomifiThemeConsumer } from '../../theme';
import { HtmlA } from '../../../reset';
import {
  BaseLinkProps,
  baseClassName,
  linkClassNames,
} from '../BaseLink/BaseLink';

export interface LinkProps extends BaseLinkProps {
  /** Ref  is passed to the anchor element. Alternative to React `ref` attribute. */
  forwardedRef?: React.Ref<HTMLAnchorElement>;
}

const StyledLink = styled(
  ({
    asProp,
    className,
    theme,
    underline = 'hover',
    ...passProps
  }: LinkProps & SuomifiThemeProp) => (
    <HtmlA
      {...passProps}
      className={classnames(baseClassName, className, {
        [linkClassNames.linkUnderline]: underline === 'initial',
      })}
      as={asProp}
    />
  ),
)`
  ${({ theme }) => LinkStyles(theme)}
`;

/**
 * <i class="semantics" />
 * Used for adding a link
 */
const Link = forwardRef(
  (props: LinkProps, ref: React.Ref<HTMLAnchorElement>) => (
    <SuomifiThemeConsumer>
      {({ suomifiTheme }) => (
        <StyledLink theme={suomifiTheme} forwardedRef={ref} {...props} />
      )}
    </SuomifiThemeConsumer>
  ),
);

Link.displayName = 'Link';
export { Link };
