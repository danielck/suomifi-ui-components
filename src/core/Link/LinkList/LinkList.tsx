import React, { forwardRef } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { LinkListStyles } from './LinkList.baseStyles';
import { SuomifiThemeProp, SuomifiThemeConsumer } from '../../theme';
import { HtmlUlProps, HtmlUlWithRef } from '../../../reset';

const LinkListClassName = 'fi-link-list';

export interface LinkListProps extends HtmlUlProps {
  /** Ref  is passed to the anchor element. Alternative to React `ref` attribute. */
  forwardedRef?: React.Ref<HTMLAnchorElement>;
}

const StyledLinkList = styled(
  ({
    className,
    theme,
    children,
    ...passProps
  }: LinkListProps & SuomifiThemeProp) => (
    <HtmlUlWithRef
      {...passProps}
      className={classnames(className, LinkListClassName)}
    >
      {children}
    </HtmlUlWithRef>
  ),
)`
  ${() => LinkListStyles()}
`;

/**
 * <i class="semantics" />
 * Used for adding a link
 */
const LinkList = forwardRef(
  (props: LinkListProps, ref: React.Ref<HTMLUListElement>) => (
    <SuomifiThemeConsumer>
      {({ suomifiTheme }) => (
        <StyledLinkList theme={suomifiTheme} forwardedRef={ref} {...props} />
      )}
    </SuomifiThemeConsumer>
  ),
);

LinkList.displayName = 'LinkList';
export { LinkList };
