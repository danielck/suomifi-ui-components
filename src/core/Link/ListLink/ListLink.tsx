import React, { forwardRef, ReactElement } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { ListLinkStyles } from './ListLink.baseStyles';
import { SuomifiThemeProp, SuomifiThemeConsumer } from '../../theme';
import { HtmlDiv, HtmlLi, HtmlSpan } from '../../../reset';
import { LinkProps } from '../Link/Link';
import { ExternalLinkProps } from 'core/Link/ExternalLink/ExternalLink';
import { RouterLinkProps } from 'core/Link/RouterLink/RouterLink';
import { Icon } from '../../Icon/Icon';
import { BaseIconKeys } from 'suomifi-icons';

const baseClassName = 'fi-list-link';

const listLinkClassNames = {
  icon: `${baseClassName}_icon`,
};

export interface ListLinkProps {
  /** Ref  is passed to the anchor element. Alternative to React `ref` attribute. */
  forwardedRef?: React.Ref<HTMLAnchorElement>;
  /**
   * Custom icon to use for the list item
   */
  icon?: BaseIconKeys;
  /**
   * The link element of the list item
   */
  children: ReactElement<
    LinkProps | ExternalLinkProps | RouterLinkProps<React.ElementType>
  >;
  className?: string;
  /** Use for list style links outside of a list element. Changes outermost element from `li` to `div` */
  standalone?: boolean;
}

const StyledListLink = styled(
  ({
    children,
    theme,
    className,
    standalone = false,
    icon,
    ...passProps
  }: ListLinkProps & SuomifiThemeProp) =>
    standalone ? (
      <HtmlDiv {...passProps} className={classnames(baseClassName, className)}>
        <HtmlSpan className={listLinkClassNames.icon}>
          {!!icon ? (
            <Icon icon={icon} />
          ) : (
            <Icon icon={'chevronRight'} color={theme.colors.highlightBase} />
          )}
        </HtmlSpan>
        {children}
      </HtmlDiv>
    ) : (
      <HtmlLi {...passProps} className={classnames(baseClassName, className)}>
        <HtmlSpan className={listLinkClassNames.icon}>
          {!!icon ? (
            <Icon icon={icon} />
          ) : (
            <Icon icon={'chevronRight'} color={theme.colors.highlightBase} />
          )}
        </HtmlSpan>
        {children}
      </HtmlLi>
    ),
)`
  ${({ theme }) => ListLinkStyles(theme)}
`;

/**
 * <i class="semantics" />
 * Used for adding a link
 */
const ListLink = forwardRef(
  (
    props: ListLinkProps,
    ref: React.Ref<HTMLAnchorElement> /* TODO: ref typing */,
  ) => (
    <SuomifiThemeConsumer>
      {({ suomifiTheme }) => (
        <StyledListLink theme={suomifiTheme} forwardedRef={ref} {...props} />
      )}
    </SuomifiThemeConsumer>
  ),
);

ListLink.displayName = 'ListLink';
export { ListLink };
