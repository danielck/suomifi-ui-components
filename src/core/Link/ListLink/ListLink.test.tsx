import React from 'react';
import { render } from '@testing-library/react';
import { axeTest } from '../../../utils/test';

import { ListLink } from './ListLink';
import { LinkList } from '../LinkList/LinkList';
import { Link } from '../Link/Link';

const TestListLink = (
  <LinkList>
    <ListLink data-testid="test-link">
      <Link href="/">Linkki</Link>
    </ListLink>
  </LinkList>
);

test('calling render with the same component on the same container does not remount', () => {
  const ListLinkRendered = render(TestListLink);
  const { getByTestId, container } = ListLinkRendered;
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('test-link').textContent).toBe('Linkki');
});

test('should not have underline class by default', () => {
  const { container } = render(TestListLink);
  expect(container.firstChild).not.toHaveClass('fi-link--initial-underline');
});

test('should not have basic accessibility issues', axeTest(TestListLink));
