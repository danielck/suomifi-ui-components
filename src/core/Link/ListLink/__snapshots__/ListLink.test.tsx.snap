// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c6 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  background-color: transparent;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c6::before,
.c6::after {
  box-sizing: border-box;
}

.c6:hover,
.c6:active,
.c6:focus,
.c6:focus-within {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: list-item;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4::before,
.c4::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  list-style-type: decimal;
  margin-top: 1em;
  margin-bottom: 1em;
  margin-left: 0;
  margin-right: 0;
  padding-left: 40px;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c5 {
  vertical-align: baseline;
}

.c5.fi-icon {
  display: inline-block;
}

.c5 .fi-icon-base-fill {
  fill: currentColor;
}

.c5 .fi-icon-base-stroke {
  stroke: currentColor;
}

.c5.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c5.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c3.fi-list-link {
  list-style: none;
}

.c3.fi-list-link .fi-list-link_icon {
  margin-right: 2px;
}

.c3.fi-list-link .fi-list-link_icon .fi-icon {
  margin-left: -3px;
  -webkit-transform: translateY(0.1em);
  -ms-transform: translateY(0.1em);
  transform: translateY(0.1em);
}

.c3.fi-list-link .fi-list-link_icon .fi-icon .fi-icon-base-fill {
  fill: hsl(23,82%,51%);
}

.c1.fi-link-list {
  list-style: none;
  padding: 0;
  margin: 5px 0 0 0;
}

.c7 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
}

.c7.fi-link {
  color: hsl(212,63%,45%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7.fi-link:hover,
.c7.fi-link:active,
.c7.fi-link:focus,
.c7.fi-link:focus-within {
  color: hsl(212,63%,45%);
}

.c7.fi-link:focus,
.c7.fi-link:focus-within {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7.fi-link:focus {
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  outline: 0;
}

.c7.fi-link:hover,
.c7.fi-link:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c7.fi-link:visited {
  color: hsl(284,36%,45%);
}

.c7.fi-link--initial-underline {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c7.fi-link--initial-underline:focus,
.c7.fi-link--initial-underline:focus-within {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c7.fi-link--initial-underline:hover,
.c7.fi-link--initial-underline:active {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<ul
  class="c0 c1 fi-link-list"
>
  <li
    class="c2 fi-list-link c3"
    data-testid="test-link"
  >
    <span
      class="c4 fi-list-link_icon"
    >
      <svg
        aria-hidden="true"
        class="fi-icon c5"
        focusable="false"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="fi-icon-base-fill"
          d="M15.61 11.01c.26.273.39.632.39.99s-.13.717-.39.99l-5.334 5.6a1.287 1.287 0 01-1.885 0 1.448 1.448 0 010-1.98l4.39-4.61-4.39-4.61a1.448 1.448 0 010-1.98 1.287 1.287 0 011.885 0l5.333 5.6z"
          fill="#222"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <a
      class="c6 fi-link c7"
      href="/"
    >
      Linkki
    </a>
  </li>
</ul>
`;
